httomo/ui_layer.py

```diff
@@ -0,0 +1,118 @@
+            self.PipelineStageConfig = _python_tasks_loader(self.tasks_file_path)
+        else:
+            raise ValueError(
+                f"The extension {ext} of the file {root} with tasks is unknown."
+            )
+
+    def build_pipeline(self) -> Pipeline:
+        side_outputs_collect: list = [] # saves [task_no, id, side_outputs] for tasks with side_outputs
+        methods_list: list = []
+        for task_no, task_conf in enumerate(self.PipelineStageConfig):
+            if "loaders" in task_conf['module_path']:
+                task_conf['parameters']['in_file'] = self.in_data_file
+                # unpack params and initiate a loader
+                loader =  make_loader(
+                        self.repo,
+                        task_conf['module_path'],
+                        task_conf['method'],
+                        self.comm,
+                        **task_conf['parameters'],
+                        )
+            else:
+                if "parameters" not in task_conf:
+                    task_conf['parameters'] = {}
+                if "side_outputs" not in task_conf:
+                    task_conf['side_outputs'] = {}
+                else:
+                    side_outputs_collect.append([task_no, task_conf['id'], task_conf['side_outputs']])
+                # check if there is a reference to side_outputs to cross-link
+                for key, value in task_conf['parameters'].items():
+                    if isinstance(value, str) and value is not None:
+                        if value.find('${{') != -1:                            
+                            result_extr = re.search(r"\{([A-Za-z0-9_.]+)\}", value)
+                            internal_expression = result_extr.group(1)
+                            (ref_id, side_str, ref_arg) = internal_expression.split(".")
+                            # lets find the referred id in "side_outputs_collect"
+                            for items in side_outputs_collect:
+                                if items[1] == ref_id:
+                                    # refer to methods_list[items[0]-1]
+                                    task_conf['parameters'][key] = OutputRef(methods_list[items[0]-1], ref_arg)
+                # unpack params of a method and append to a list of methods
+                method = make_backend_wrapper(
+                    self.repo,
+                    task_conf['module_path'],
+                    task_conf['method'],
+                    self.comm,
+                    task_conf['side_outputs'],
+                    **task_conf['parameters'],
+                )
+                methods_list.append(method)
+        return Pipeline(
+            loader=loader,
+            methods=methods_list,
+        )
+
+def _yaml_loader(file_path: str) -> list:
+    with open(file_path, "r") as f:
+        tasks_list = list(yaml.load_all(f, Loader=yaml.FullLoader))
+    return tasks_list
```

Comment:
thanks, but this actually breaks the test and specifically the unpacking
`for task_no, task_conf in enumerate(self.PipelineStageConfig):` loop.

Reply:
Ah right, I wouldn't be able to tell if this was the cause for the error, but the `Loader` param also needs to be changed to `Loader=yaml.Loader` to get back a list instead of a generator. I forgot that the `yaml.load_all()` is handling the multliple documents in the YAML, but the `Loader` param also plays a role in the data structure that's returned, my bad.